magenta-lib/src/main/scala/magenta/deployment_type/CloudFormation.scala

```diff
@@ -44,44 +49,72 @@ object CloudFormation extends DeploymentType with CloudFormationDeploymentTypePa
         |templateParameters parameters, with stage-specific values overriding general parameters
         |when in conflict.""".stripMargin
   ).default(Map.empty)
+
   val createStackIfAbsent = Param[Boolean]("createStackIfAbsent",
     documentation = "If set to true then the cloudformation stack will be created if it doesn't already exist"
   ).default(true)
 
+  val secondsToWaitForChangeSetCreation = Param("secondsToWaitForChangeSetCreation",
+    "Number of seconds to wait for the change set to be created").default(15 * 60)
+
   val updateStack = Action("updateStack",
     """
       |Apply the specified template to a cloudformation stack. This action runs an asynchronous update task and then
       |runs another task that _tails_ the stack update events (as well as possible).
       |
     """.stripMargin
   ){ (pkg, resources, target) => {
-      implicit val keyRing = resources.assembleKeyring(target, pkg)
-      implicit val artifactClient = resources.artifactClient
-      val reporter = resources.reporter
-
-      val amiParameterMap: Map[CfnParam, TagCriteria] = getAmiParameterMap(pkg, target, reporter)
-
-      val cloudFormationStackLookupStrategy = getCloudFormationStackLookupStrategy(pkg, target, reporter)
-
-      val globalParams = templateParameters(pkg, target, reporter)
-      val stageParams = templateStageParameters(pkg, target, reporter).lift.apply(target.parameters.stage.name).getOrElse(Map())
-      val params = globalParams ++ stageParams
-      List(
-        UpdateCloudFormationTask(
-          target.region,
-          cloudFormationStackLookupStrategy,
-          S3Path(pkg.s3Package, templatePath(pkg, target, reporter)),
-          params,
-          amiParameterMap,
-          getLatestAmi(pkg, target, reporter, resources.lookup),
-          target.parameters.stage,
-          target.stack,
-          createStackIfAbsent(pkg, target, reporter),
-          true
-        ),
-        CheckUpdateEventsTask(target.region, cloudFormationStackLookupStrategy)
-      )
+    implicit val keyRing = resources.assembleKeyring(target, pkg)
+    implicit val artifactClient = resources.artifactClient
+    val reporter = resources.reporter
+
+    val amiParameterMap: Map[CfnParam, TagCriteria] = getAmiParameterMap(pkg, target, reporter)
+    val cloudFormationStackLookupStrategy = getCloudFormationStackLookupStrategy(pkg, target, reporter)
+
+    val globalParams = templateParameters(pkg, target, reporter)
+    val stageParams = templateStageParameters(pkg, target, reporter).lift.apply(target.parameters.stage.name).getOrElse(Map())
+
+    val userParams = globalParams ++ stageParams
+    val amiLookupFn = getLatestAmi(pkg, target, reporter, resources.lookup)
+
+    val stackTags = cloudFormationStackLookupStrategy match {
+      case LookupByTags(tags) => Some(tags)
+      case _ => None
     }
+
+    val changeSetName = s"riff-raff-${UUID.randomUUID().toString}"
```

Comment:
> I imagine the concern is that it could fail if it exists?
Yep, all changesets for a CFN stack must have unique names. Is there a unique identifier for each deploy we can access from the params?

Reply:
Unique even after it is deleted?

Reply:
Ah, I forgot we do that. I tested it in the console and you can indeed re-use names after deletion.

Reply:
It would be better I think if it were something like `riff-raff-$escapedProjectName-$stage-$buildId-$deploymentName-$region`. Perhaps you don't need region as it will already be unique to that.